### PR TITLE
fix: preserve metadata in proto

### LINKED ISSUE #1977: DFSchema serialization does not preserve metadata
**Describe the bug**
As discovered on https://github.com/apache/arrow-datafusion/pull/1914 https://github.com/apache/arrow-datafusion/pull/1914#issuecomment-1063984916, when a `DFSchema` is serialized, any name=value metadata is lost

**To Reproduce**
1. Try and serialize a DF schema with metadata -- when it is deserialized the metadata will be lost

**Expected behavior**
1. Remove the `#[allow(deprecated)]` added in #1914

### PR DESCRIPTION
**Pull request by xudong963**
# Which issue does this PR close?

Closes #1977 

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Review by houqp — APPROVED** [review]

File: datafusion-proto/src/from_proto.rs [hunk 1]
```diff
@@ -155,9 +156,12 @@ impl TryFrom<&protobuf::DfSchema> for DFSchema {
...
             .collect::<Result<Vec<DFField>, _>>()?;
-        // https://github.com/apache/arrow-datafusion/issues/1977
-        #[allow(deprecated)]
-        Ok(DFSchema::new(fields)?)
+        let metadata = df_schema
+            .metadata
+            .iter()
+            .map(|m| (m.0.clone(), m.1.clone()))
+            .collect::<HashMap<String, String>>();
+        Ok(DFSchema::new_with_metadata(fields, metadata)?)
```
Comment (yjshen) [inline datafusion-proto/src/from_proto.rs:164]:
```suggestion
        Ok(DFSchema::new_with_metadata(fields, df_schema.metadata.clone())?)
```

File: datafusion-proto/src/to_proto.rs [hunk 2]
```diff
@@ -290,7 +291,12 @@ impl From<&DFField> for protobuf::DfField {
 impl From<&DFSchemaRef> for protobuf::DfSchema {
     fn from(s: &DFSchemaRef) -> protobuf::DfSchema {
         let columns = s.fields().iter().map(|f| f.into()).collect::<Vec<_>>();
-        protobuf::DfSchema { columns }
+        let metadata = s
+            .metadata()
+            .iter()
+            .map(|m| (m.0.clone(), m.1.clone()))
+            .collect::<HashMap<_, _>>();
+        protobuf::DfSchema { columns, metadata }
```
Comment (yjshen) [inline datafusion-proto/src/to_proto.rs:299]:
```suggestion
        protobuf::DfSchema { columns , metadata: s.metadata().clone() }
```

**Review by yjshen — COMMENTED** [review]
Seems we could replace with `clone`?

**Review by liukun4515 — APPROVED** [review]
LGTM

**Comment by yjshen** [comment]
Thanks @xudong963 !

**Pull request merged**